### PR TITLE
Memory usage improvements

### PR DESCRIPTION
**Pull request by alexmaco**
These are mostly focused on splitting up enum variants with large size differences between variants by `Box`-ing things up.

In my testing this reduces the memory usage somewhere in the low percentages, even though the measurements are quite noisy.

### CODE REVIEW

**Comment by lnicola** [comment]
> In my testing this reduces the memory usage somewhere in the low percentages, even though the measurements are quite noisy.

Did you try `rust-analyzer analysis-stats --memory-usage`? It's quite stable on CI: https://rust-analyzer.github.io/metrics/

File: crates/flycheck/src/lib.rs [hunk 1]
```diff
@@ -128,7 +128,7 @@ struct FlycheckActor {
 
 enum Event {
     Restart(Restart),
-    CheckEvent(Option<cargo_metadata::Message>),
+    CheckEvent(Option<Box<cargo_metadata::Message>>),
```
Comment (bjorn3) [inline crates/flycheck/src/lib.rs:131]: This is the most common variant, so using a `Box` here wil actually increase the  average memory usage.

File: crates/vfs/Cargo.toml [hunk 2]
```diff
@@ -14,3 +14,4 @@ rustc-hash = "1.0"
 fst = "0.4"
 
 paths = { path = "../paths", version = "0.0.0" }
+indexmap = "1.6.2"
```
Comment (bjorn3) [inline crates/vfs/Cargo.toml:17]: nit: missing trailing newline

File: crates/vfs/src/path_interner.rs [hunk 3]
```diff
@@ -1,37 +1,42 @@
...
 
     /// Insert `path` in `self`.
     ///
     /// - If `path` already exists in `self`, returns its associated id;
     /// - Else, returns a newly allocated id.
     pub(crate) fn intern(&mut self, path: VfsPath) -> FileId {
-        if let Some(id) = self.get(&path) {
-            return id;
+        let (id, added) = self.map.insert_full(path);
+        if added {
```
Comment (bjorn3) [inline crates/vfs/src/path_interner.rs:36]: Unconditionally doing the assert could be faster.

File: crates/flycheck/src/lib.rs (hunk 1 above)
Comment (alexmaco) [reply crates/flycheck/src/lib.rs]: removed this commit

File: crates/vfs/src/path_interner.rs (hunk 3 above)
Comment (alexmaco) [reply crates/vfs/src/path_interner.rs]: removed condition

**Comment by alexmaco** [comment]
> Did you try `rust-analyzer analysis-stats --memory-usage path/to/some/repo`? It's quite stable on CI (look at the second graph): https://rust-analyzer.github.io/metrics/. Unfortunately, it only works on Linux IIRC.

That looks indeed pretty stable. On rust-analyzer this PR improves memory usage by 4.3%. The improvements on other projects are in the same ballpark, between 1.8% and ~4%.

**Comment by jonas-schievink** [comment]
Nice!

bors r+

**Comment by lnicola** [comment]
This showed up in the metrics:

image

**Comment by lnicola** [comment]
changelog internal

**Pull request merged**